### ISSUE TITLE
"no-unused-vars" doesn't work with recursive function correctly, when it is declared with function expression

### ISSUE THREAD

**Issue by aliaksandr-yermalayeu** [eslint/eslint#10982]
**Tell us about your environment**

* **ESLint Version:** 5.7.0
* **Node Version:** 8.11.1
* **npm Version:** 5.6.0

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
```
{
    "rules": {
        "no-unused-vars": "error"
    }
}
```
<details>
<summary>Configuration</summary>

```json
{
    "rules": {
        "no-unused-vars": "error"
    }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

_function declaration_
```js
// considered error
function a () {
    a();
}
```

_function expression_
```js
// error is not happened
var a = function () {
    a();
}
```

**What did you expect to happen?**
Second case should cause error the same way as the first one.

**Comment by platinumazure** [comment]
Hi @aliaksandr-yermalayeu, thanks for the issue.

It seems to me the recursive declaration case should be flagged as used (i.e., neither example should produce a warning). But I'm not 100% sure.

@not-an-aardvark @mysticatea Any thoughts?

**Comment by not-an-aardvark** [comment]
The firat example should produce a warning because the function can never be initially called, since it's only referenced from inside itself.

The second example could be made to produce a warning for the same reason, although this would require modifying the rule to be able to infer that the function can only be accessed through the `a` variable.

**Comment by mysticatea** [comment]
It has been introduced in #2095; the rule is addressing function declarations to detect recursive calls, but not variable declarations. We can enhance the rule. Maybe...

- the rule should ignore references in the `VariableDeclarator#init` and `AssignmentPattern#right` if duplicate variables don't exist.

**Comment by kaicataldo** [comment]
@mysticatea Just to clarify, are you saying this is a bug?

**Comment by mysticatea** [comment]
Looks like an enhancement to me.
